common/crypto/platform/openssl/crypto.cpp

```diff
@@ -139,6 +143,97 @@ ExpectedPrivateKey PrivateKey::LoadFromPEM(const string &private_key_path) {
+#ifdef MENDER_CRYPTO_OPENSSL_LEGACY
+	auto pkey_gen_ctx = unique_ptr<EVP_PKEY_CTX, void (*)(EVP_PKEY_CTX *)>(
+		EVP_PKEY_CTX_new_id(EVP_PKEY_RSA, nullptr), pkey_ctx_free_func);
+
+	int ret = EVP_PKEY_keygen_init(pkey_gen_ctx.get());
+	if (ret != OPENSSL_SUCCESS) {
+		return expected::unexpected(MakeError(
+			SetupError,
+			"Failed to generate a private key. Initialization failed: "
+				+ GetOpenSSLErrorMessage()));
+	}
+
+	ret = EVP_PKEY_CTX_set_rsa_keygen_bits(pkey_gen_ctx.get(), bits);
+	if (ret != OPENSSL_SUCCESS) {
+		return expected::unexpected(MakeError(
+			SetupError,
+			"Failed to generate a private key. Parameters setting failed: "
+				+ GetOpenSSLErrorMessage()));
+	}
+
+	auto exponent_bn = unique_ptr<BIGNUM, void (*)(BIGNUM *)>(BN_new(), [](BIGNUM *bn) {
+		// no-op; the ownership is passed to EVP_PKEY_CTX_set_rsa_keygen_pubexp
+	});
+	ret = BN_set_word(exponent_bn.get(), exponent);
+	if (ret != OPENSSL_SUCCESS) {
+		return expected::unexpected(MakeError(
+			SetupError,
+			"Failed to generate a private key. Parameters setting failed: "
+				+ GetOpenSSLErrorMessage()));
+	}
+
+	ret = EVP_PKEY_CTX_set_rsa_keygen_pubexp(pkey_gen_ctx.get(), exponent_bn.get());
+	if (ret != OPENSSL_SUCCESS) {
+		return expected::unexpected(MakeError(
+			SetupError,
+			"Failed to generate a private key. Parameters setting failed: "
+				+ GetOpenSSLErrorMessage()));
+	}
+
+	EVP_PKEY *pkey = nullptr;
+	ret = EVP_PKEY_keygen(pkey_gen_ctx.get(), &pkey);
+	if (ret != OPENSSL_SUCCESS) {
+		return expected::unexpected(MakeError(
+			SetupError,
+			"Failed to generate a private key. Generation failed: " + GetOpenSSLErrorMessage()));
+	}
+#else
+	auto pkey_gen_ctx = unique_ptr<EVP_PKEY_CTX, void (*)(EVP_PKEY_CTX *)>(
+		EVP_PKEY_CTX_new_from_name(NULL, "RSA", NULL), pkey_ctx_free_func);
+
+	int ret = EVP_PKEY_keygen_init(pkey_gen_ctx.get());
+	if (ret != OPENSSL_SUCCESS) {
+		return expected::unexpected(MakeError(
+			SetupError,
+			"Failed to generate a private key. Initialization failed: "
+				+ GetOpenSSLErrorMessage()));
+	}
+
+	OSSL_PARAM params[3];
+	auto bits_buffer = bits;
+	auto exponent_buffer = exponent;
+	params[0] = OSSL_PARAM_construct_uint("bits", &bits_buffer);
+	params[1] = OSSL_PARAM_construct_uint("e", &exponent_buffer);
+	params[2] = OSSL_PARAM_construct_end();
+
+	ret = EVP_PKEY_CTX_set_params(pkey_gen_ctx.get(), params);
+	if (ret != OPENSSL_SUCCESS) {
+		return expected::unexpected(MakeError(
+			SetupError,
+			"Failed to generate a private key. Parameters setting failed: "
+				+ GetOpenSSLErrorMessage()));
+	}
+
+	EVP_PKEY *pkey = nullptr;
+	ret = EVP_PKEY_generate(pkey_gen_ctx.get(), &pkey);
+	if (ret != OPENSSL_SUCCESS) {
+		return expected::unexpected(MakeError(
+			SetupError,
+			"Failed to generate a private key. Generation failed: " + GetOpenSSLErrorMessage()));
+	}
+#endif
+
+	auto private_key = unique_ptr<EVP_PKEY, void (*)(EVP_PKEY *)>(pkey, pkey_free_func);
```

Comment:
I know this is pre-existing, but there is something weird with our interface don't you think? :thinking:

Reply:
Can you elaborate?

Reply:
The `PrivateKey` class interface strikes me as a little odd.

It takes a `unique_ptr` and has to be wrapped in a `unique_ptr` itself

Reply:
Yeah, I see now. I will let @larsewi take a value add day to fix this :sweat_smile: